### PR TITLE
[opensprinkler] Make http connection more resilient

### PR DESCRIPTION
**Pull request by wzbfyb**
**Improvement**

-     connection timeout is now configurable
-     on connection timeout, retries are possible and configurable
-     retry delay is configurable

**Reason**
my OS has a flaky WiFi connection and sometimes goes offline due to timeouts.

### CODE REVIEW

**Comment by wzbfyb** [comment]
sorry for the mess with multiple pull requests, I made a mistake when initially forking the code.

**Review by Hilbrand — COMMENTED** [review]
Also update the readme to include the new parameters.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java [hunk 1]
```diff
@@ -379,22 +387,44 @@ public String sendHttpGet(String url, @Nullable String urlParameters)
...
+                try {
+                    response = withGeneralProperties(httpClient.newRequest(location))
+                            .timeout(config.timeout, TimeUnit.SECONDS).method(HttpMethod.GET).send();
+                    connectionSuccess = true;
+                } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                    // throw new CommunicationApiException("Request to OpenSprinkler device failed: " + e.getMessage());
+                    logger.warn("Request to OpenSprinkler device failed (retries left: {}): {}", retriesLeft,
+                            e.getMessage());
+                    try {
+                        Thread.sleep(1000 * config.retryDelay);
```
[the remark this answers is not in the document]
Comment (wzbfyb) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java:404]: Indeed. As it is not necessary to achieve the desired effect, I will just get rid of the delay. Working with appropriate timeouts should suffice.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/config/OpenSprinklerHttpInterfaceConfig.java [hunk 2]
```diff
@@ -43,6 +43,22 @@ public class OpenSprinklerHttpInterfaceConfig {
...
+
+    /**
+     * Number of retries in case of connection timeouts
+     */
+    public int retry = 3;
+
+    /**
+     * Number of seconds delay between connection retries
+     */
+    public int retryDelay = 10;
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/config/OpenSprinklerHttpInterfaceConfig.java:60]: The unused field `retryDelay` can be removed.

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -32,6 +32,18 @@
 				<description>Specifies the refresh interval in seconds.</description>
 				<default>60</default>
 			</parameter>
+			<parameter name="retry" type="integer">
+				<label>Retry Count</label>
+				<description>Specifies the number of retries on connection timeouts.</description>
+				<default>3</default>
+				<advanced>true</advanced>
+			</parameter>
+			<parameter name="timeout" type="integer">
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:41]: I forgot to mention. But you can also add a unit to parameters.
```suggestion
			<parameter name="timeout" type="integer" unit="s">
```

**Comment by lolodomo** [comment]
Was it finally ok for you @Hilbrand  ?

**Comment by jlaur** [comment]
Gentle ping @Hilbrand 🙂

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -32,6 +32,18 @@
 				<description>Specifies the refresh interval in seconds.</description>
 				<default>60</default>
 			</parameter>
+			<parameter name="retry" type="integer">
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:35]: Please update the i18n properties file to add these 2 new parameters.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Just one comment about i18n.

**Comment by Skinah** [comment]
Just in case the author @wzbfyb does not know how to do the last step to get this merged, please see the mvn command you need as well as other docs here:
https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

**Comment by wzbfyb** [comment]
hi!

sorry, missed the notification. will take a look in the next days.
thanks a lot!



------ Originalnachricht ------
Von "Matthew Skinner" ***@***.***>
An "openhab/openhab-addons" ***@***.***>
Cc "wzbfyb" ***@***.***>; "Mention" 
***@***.***>
Datum 10.12.23, 02:55:32
Betreff Re: [openhab/openhab-addons] [opensprinkler] Make http 
connection more resilient (PR #14998)

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml (hunk 4 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:35]: To do it simply, run: mvn i18n:generate-default-tramslations

Comment (lolodomo) [reply bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:35]: To accelerate process and merge this PR before OH 4.1 is released, I run the command myself and push the change.

**Review by lolodomo — APPROVED** [review]

**Comment by wzbfyb** [comment]
Thanks a lot, highly appreciated.


------ Originalnachricht ------
Von "lolodomo" ***@***.***>
An "openhab/openhab-addons" ***@***.***>
Cc "wzbfyb" ***@***.***>; "Mention" 
***@***.***>
Datum 14.12.2023 21:55:24
Betreff Re: [openhab/openhab-addons] [opensprinkler] Make http 
connection more resilient (PR #14998)

**Pull request merged**